platoon/channel/worker.py

```diff
@@ -0,0 +1,569 @@
+            This method used to be called `lock_params`.
+
+        """
+        self._lock.acquire(timeout)
+
+    def unlock(self):
+        """
+        Release the lock across all workers.
+
+        The current implementation does not ensure that the process
+        that locked the params is the one that unlocks them. It also
+        does not prevent one process from unlocking more than once
+        (which will allow more than one process to hold the lock). Adds 1 in
+        underlying semaphore.
+
+        Make sure you follow proper lock/unlock logic in your program
+        to avoid these problems.
+
+        .. versionchanged:: 0.6.0
+            This method used to be called `unlock_params`.
+
+        """
+        self._lock.release()
+
+    @property
+    def local_size(self):
+        "Number of workers assigned to local host's controller"
+        return self._local_size
+
+    @property
+    def global_size(self):
+        "Number of workers spawned across all hosts in total"
+        return self._global_size
+
+################################################################################
+#                   Initialization and Finilization Methods                    #
+################################################################################
+
+    def _handle_force_close(self, signum, frame):
+        """Handle SIGINT signals from Controller.
+
+        This is expected to happen when something abnormal has happened in other
+        workers which implies that training procedure should stop and fail.
+
+        """
+        self.close()
+        sys.exit(1)  # Exit normally with non success value.
+
+    def close(self):
+        print("Closing connections and unlinking memory...", file=sys.stderr)
+        if hasattr(self, 'asocket'):
+            self.asocket.close()
+        if hasattr(self, 'csocket'):
+            self.csocket.close()
+        self.context.term()
+        self._lock.close()
+        try:
+            self._lock.unlink()
+        except posix_ipc.ExistentialError:
+            pass
+        if hasattr(self, '_shmref'):
+            try:
+                self._shmref.unlink()
+            except posix_ipc.ExistentialError:
+                pass
+        for shmref in self._shmrefs.values():
+            try:
+                shmref.unlink()
+            except posix_ipc.ExistentialError:
+                pass
+
+    def _register_to_platoon(self):
+        if pygpu:
+            self.ctx_name = None
+            self.gpuctx = theanoga.get_context(self.ctx_name)
+            self.device = theanoconf.device
+            self._local_id = gpucoll.GpuCommCliqueId(context=self.gpuctx)
+            # Ask controller for local's info to participate in
+            response = self.send_req("platoon-get_platoon_info",
+                                     info={'device': self.device,
+                                           'local_id': self._local_id.comm_id.decode('utf-8')})
```

Comment:
Unfortunately, although I don't understand why is that... we cannot. We will get the TypeError: "bytearray(b'') is not JSON serializable".

https://bugs.python.org/issue10976
https://stackoverflow.com/questions/26862437/bytearray-is-not-json-serializable
https://docs.python.org/2/library/json.html#encoders-and-decoders